stdlib/trains/trains.lua

```diff
@@ -0,0 +1,195 @@
+--- Trains module
+-- <p>When this module is loaded, it registers new
+-- events in order to track trains as locomotives and
+-- carriages are moved around. For this reason, you should use
+-- this library's <a href="Event.html">Event</a> module</p>
+-- @module Trains
+
+-- Event registration is performed at the bottom of this file,
+-- once all other functions have been defined
+
+require 'stdlib/event/event'
+require 'stdlib/surface'
+require 'stdlib/table'
+
+local registry_loaded = false
+
+local Trains = {
+    _registry = {}
+}
+
+--- Event id for the custom event fired
+-- when a train has been renamed
+-- @usage
+----Event.register(Trains.on_train_renamed, my_handler)
+Trains.on_train_renamed = script.generate_event_name()
+
+--- Given search criteria (a table that contains at least a surface_name)
+-- searches the given surface for trains that match the criteria
+-- @usage
+----Trains.find_filtered({ surface_name = "nauvis", state = defines.train_state.wait_station })
+-- @tparam Table criteria A table with at least the surface_name key
+--  <p>
+--  Options:
+--  <ol>
+--      <li>surface_name</li>
+--      <li>state</li>
+--  </ol>
+-- </p>
+-- @return A list of train details tables, if any are found matching the criteria. Otherwise the empty list.
+-- <p>
+--  <table>
+--      <tr><td>train (LuaTrain)</td><td>The LuaTrain instance</td></tr>
+--      <tr><td>id (int)</td><td>The id of the train</td></tr>
+--  </table>
+-- </p>
+function Trains.find_filtered(criteria)
+    -- Get locomotives by filter
+    surface_index = criteria.surface_name or 1
+    locomotives = Surface.find_all_entities({

```

Comment:
I'm a bit confused here. The only criteria you've given to `Surface.find_all_entities` is the surface index. Won't this return *all* entities, not just trains? You might want to tweak your tests to add a non-train entity and make sure this is working as you expect it to...

I think you need to qualify the type of entity.